docs/src/proposals/vote_subcommittee.md

```diff
@@ -0,0 +1,158 @@
+sha-ni hardware.
+
+* secondary rotation: when a secondary subcommittee is switched
+
+* primary rotation: when primary and secondary subcommittees switch
+
+### Safety Violations
+
+An Optimistic Confirmation violation occurs if a voting subcommittee
+optimistically confirms a slot that wasn't rooted by the following
+voting subcommittee.
+
+A Safety violation occurs if a subcommittee roots a slot that wasn't
+rooted by the following subcommittee.
+
+All full nodes including RPCs should halt as soon as a safety
+violation is detected.
+
+### Subcommittee Rotation
+
+```
+
+a0 a1 A1 A1 a1 a2 A2 A2 a2 a3
+B1 B1 b1 b2 B2 B2 b2 b3 B3 B3
+```
+
+To rotate, the subcommittee must go through the rotation protocol.
+In the above example, lowercase letter represents **secondary**
+while uppercase represents **primary**.
+
+Each transition like (A1,b1) -> (A1, b2) occurs on blocks that cross
+the epoch boundary and are descendants of a rooted block from the
+previous epoch.
+
+The transition starts at a bank that crosses the epoch boundary.
+If it is a descendant of a rooted bank in the previous epoch then
+the transition is active.  If the block is not a descendant, then
+the transition is inactive.
+
+There are two types of transitions
+
+**primary rotation**: (a1,B1) -> (A1, b1)
+* active blocks: A1 is the primary and fork weight follows A1
+* inactive blocks: B1 is the primary and fork weight follows B1
+
+During the **primary rotation** the subcommittees remain constant
+but change flip their **primary/secondary** position.
+
+**secondary rotation**: (A1,b1) -> (A1, b2)
+* active blocks: A1 is the primary and fork weight follows A1
+* inactive blocks: A1 is the primary and fork weight follows A1
+
+#### Primary Rotation
+
+During the **primary rotation** two possible forks can occur.
+
+* Epoch 0: (a0, B1) - ALL epoch 2 blocks have a root to this epoch
+* Epoch 1: (a1, B1) - SOME epoch 2 blocks have a root to this epoch
+* Epoch 2: (A1, b1)
+
+For the transition to be complete, (A1, b1), must root a block
+together in Epoch 1. Therefore block producers should continue
+picking **B1/b1** during epoch 2.
+
+For blocks propsed in epoch 2, **A1** may use **B1/b1**'s votes
+from epoch 1 and 2 for switching proofs.
+
+For blocks propsed in epoch 2, all block producers should be following
+B1/b1 for fork weigt.
+
+Optimistic confirmation is valid iff both vote 2/3+ on the same fork.
+
+* Epoch 0: (a0, B1) - ALL epoch 2 blocks have a root to this epoch
+* Epoch 1: (a1, B1) - SOME epoch 2 blocks have a root to this epoch
+* Epoch 2: (A1, b1)
+* Epoch 3: (A1, b1|b2) - SOME blocks may have b1 or b2 votes
+
+If there is a root in epoch 2, then the **secondary rotation** will
+be in effect for those blocks, and **b2** will be the valid
+subcommittee. Block producers should stlll be following **b1** fork
+weight, so eventually **b1** will observe a rooted block in epoch 1
+with 2/3+ votes and will stop voting.
```

Comment:
This section is a bit confusing. Are we trying to prove optimistic confirmation safety? If so would be to explicitly say that first.

"If there is a root in epoch 2". What if there's not a root? We need to cover all cases, so it would be helpful to list all these cases and why we maintain safety under each one